### PR TITLE
Initialization, Master Controller, and Garbage Collector Overhaul

### LINKED ISSUE #1307: Dev: Opening the Criminal Sentencing Computer's UI causes client-side errors
Interacting with the Criminal Sentencing Computer in Security Processing prints the following messages to the DreamSeeker client log:
```
winset: Element crim_sentence not found.
winset: Element crim_sentence not found.
```
revision: 70ad90663b1240dbf6f2d19d7b80392a510c6b98

### LINKED ISSUE #1686: Lighting overlays are not correctly cleaned up on turf destruction
A turf being destroyed by an explosion results in the attached lighting overlay being placed onto a space turf, and later destroying itself when it's updated.

Lighting overlays should be cleared after their associated turf is destroyed.

### PR DESCRIPTION
**Pull request by Lohikar**
Just a few refactors.

This PR closes #1951.
Fixes #1307.
Fixes #1686.

changes:
- Ported /tg/'s StonedMC (SMC) master controller, which handles load better than PSched does as well as allowing for processes/subsystems that tick in the lobby.
- Ported /tg/'s addtimer() implementation.
- Removed `scheduler` process as it is obsoleted by addtimer.
- Ported /tg/'s garbage collector.
- Split machinery into two processes; machinery and pipenet.
- Players can now join much earlier in the initialization process.
- Made the plants controller a process/subsystem instead of its own mini-scheduler.
- `stoplag()` (used by `CHECK_TICK`) now dynamically adjusts its delay based on how much load the master controller is reporting.
- `CHECK_TICK` calls from a proc called by a subsystem will use that subsystem's tick limits instead of the global limit.
- Airlock autoclose now uses `spawn()` for most close delays, unless the door is obstructed (which uses addtimer). (Backported to development)
- Orbits now use a subsystem.
- Tesla tweaks.
- Singularity is now lag-checked.
- Disposals promoted to a subsystem - can handle hundreds of objects without breaking 6ms tick-time.
- Converted some `New()` procs to `Initialize()`.
- Killed and replaced turf processor.
- Killed some `process()` calls, replaced with timers.
- New-format effects can now get the amount of time elapsed since they last fired.
- Singularities and child-types of singularity have been moved into their own processor to allow for better MC scheduling.
- Tesla beams are now instant instead of bouncing between objects over time.
- The sun process/subsystem is now lag-checked.
- Fixed some GC failures.
- Removed datum pool calls, although the datum pool still exists.
- Added SSoverlay; provides automatic overlay caching, overlay queuing, and priority (non-cuttable) overlays.
- Added generic icon smoothing system, ported from /tg/.
- Replaced Bay's generic list sorting system with /tg/'s more flexible and speedy generic list sorting system. (Prerequisite of SMC & SStimer)
- Refactored conveyor belts to use the wireless system instead of iterating through global lists.
- Ported /tg/'s unified init system that works for all `/atom` instances & can completely replace New() - preserves arguments to New and passes to Initialize().
- Fixed paper's icon updating.
- Moved turf setup to Initialize() - significantly faster server bootup.
- Added lagchecking to ZAS.
- Refactored airlocks to use timers instead of processing.
- Added lagchecks to biogenerator & `storage/bag/`.
- Added an update_icon() queue processor, to replace APC's update throttling. Currently used by APCs, fire alarms, and air alarms.
- The changelog no longer force-sends assets to clients every time it is opened.
- NanoUI no longer force-sends assets to clients on connect.
- Browser datums no longer force-send assets to clients every time they are opened.
- `/turf/Destroy()` is now called when a turf is replaced by ChangeTurf().
- Lighting's changeturf hook has been split from the rest of changeturf for readability's sake.
- Lighting is back to using cheap visibility updates on changeturf instead of the more expensive forced updates. Appears to work most of the time.
- Space starlight updates in ChangeTurf have been moved into the lighting ChangeTurf hook and now use the faster `RANGE_TURFS` macro.

todo:
- [x] ~~Explosions process requires a rewrite for SMC.~~ Duct-taped old processor to a subsystem and it somehow works. Sufficient for now.
- [x] Make travis unit testing work.
- [x] Block observers from observing the station until init completes - server runtimes if they observe too early in init.
- [x] Fix powernets again.
- [x] Fix automatic night lighting.
- [x] Check that diseases work.
- [x] Check that chemistry works.
- [x] Fix respawn verb runtiming if run before round-start.
- [x] Silence warnings about the failsafe not existing until it has actually been initialized.
- [x] Fix centcomm window spawners.
- [x] Make sure the GC and the datum pool aren't fighting.
- [x] qdel loop in gun offhands.
- [x] Fix global announcer (ERT one known broken - seems related to runtimes in lighting?) - Issue from development.
- [x] Check power output of tweaked tesla.
- [x] Figure out what the fuck is sleeping in disposal holders and breaking SSdisposals.
- [x] Figure out why the fuck disposals are trying to get is_plating of non-turfs.
- [x] ~~Global iterators.~~ Fuck global iterators.
- [x] Fix excessive disposal outlet ejection speed.
- [x] Fix pipenet rebuild.
- [x] Re-add ZAS reboot verb.
- [x] Finish SSairflow.
- [x] OKAY WHAT THE FUCK IS SLEEPING IN DISPOSALS
- [x] Fix SSicon_smooth's baseturf.
- [x] Changelog.
- [x] Fix SSlighting pre-bake logging.

Known issues:
- [x] ZAS pressure equalization is broken.
- [x] RIGs spawned by random cargo runtime on misconfigure. (likely initialize() issue.)
- [x] Decals remain after a tile is broken down to plating.
- [x] The global area list is not initializing again.
- [x] Atmos connectors need to be re-connected to their pipe to function.
- [x] Something is sleeping in machinery - Only seems to happen for a little after roundstart.
- [x] Emagged doors still autoclose.
- [x] Body scan consoles runtime on `Destroy()` when cultified.
- [ ] NanoUIs attached to `/obj/machinery/atmospherics/omni/filter` runtime due to null parent - probably not cleaned up correctly on `Destroy()`.
- [x] Floor unwet timers are somehow firing on space and runtiming.
- [x] Ambrosia runtimes on `Initialize()`.
- [x] Gas overlays are not properly removed when an area is exposed to space.
- [x] Blood overlays interfere with overlays added to items via. SSoverlay.
- [x] Assembly cooldown does not work.

### CODE REVIEW

**Comment by NanakoAC** [comment]
i like what you're doing here, although this kind of seems like rolling too many unrelated things into one project.

addtimer, sparks, doors, cascade, etc. All of these seem fairly unrelated, maybe break them into smaller PRs?

**Comment by Lohikar** [comment]
Addtimer is dependent on the GC and SMC changes, and a lot of the other changes are dependent on addtimer.

The cascade and the sparks changes are the only ones I could really backport.

File: code/controllers/subsystems/explosives-recursive.dm [hunk 1]
```diff
@@ -5,18 +5,111 @@ var/datum/subsystem/explosives_recursive/SSkaboom
...
+		current_power = power
+		processing_turfs += start
+
+		while (processing_turfs.len)
+			var/turf/T = processing_turfs[processing_turfs.len]
+			processing_turfs.len--
+
+			// If the power won't make a difference, just skip the turf.
+			if (current_power <= 0 || QDELETED(T))
+				continue
```
Comment (Lohikar) [inline code/controllers/subsystems/explosives-recursive.dm]: Probably could break if power hits 0, it's not going to go up.

**Comment by skull132** [comment]
MEMES.

**Comment by Lohikar** [comment]
Replaced by #1958.

**Pull request closed without merging**